server/PathServices/SignalRRealtimeDataRepository.cs

```diff
@@ -0,0 +1,214 @@
+        /// <returns>A collection of arriving trains.</returns>
+        public Task<IEnumerable<RealtimeData>> GetRealtimeData(Station station)
+        {
+            return Task.FromResult(this.GetRealtimeData(station, RouteDirection.ToNY).Union(this.GetRealtimeData(station, RouteDirection.ToNJ)).Where(data => data.DataExpiration > DateTime.UtcNow));
+        }
+
+        private IEnumerable<RealtimeData> GetRealtimeData(Station station, RouteDirection direction)
+        {
+            return this.realtimeData.GetValueOrDefault((station, direction), new List<RealtimeData>());
+        }
+
+        private void PathSqlDbUpdated(object sender, EventArgs args)
+        {
+            Log.Logger.Here().Information("Recreating SignalR hubs following a PATH DB update...");
+            Task.Run(this.CreateHubConnections).Wait();
+        }
+
+        private async Task CreateHubConnections()
+        {
+            await this.CloseExistingHubConnections();
+
+            var tokenBrokerUrl = Decryption.Decrypt(await this.pathDataRepository.GetTokenBrokerUrl());
+            var tokenValue = Decryption.Decrypt(await this.pathDataRepository.GetTokenValue());
+
+            await Task.WhenAll(StationMappings.StationToSignalRTokenName.SelectMany(station =>
+                RouteDirectionMappings.RouteDirectionToDirectionKey.Select(direction => this.CreateHubConnection(tokenBrokerUrl, tokenValue, station.Key, direction.Key))));
+        }
+
+        private async Task CreateHubConnection(string tokenBrokerUrl, string tokenValue, Station station, RouteDirection direction)
+        {
+            SignalRToken token;
+
+            try
+            {
+                token = await this.pathApiClient.GetToken(tokenBrokerUrl, tokenValue, station, direction);
+
+                var connection = new HubConnectionBuilder()
+                    .WithUrl(token.Url, c => c.AccessTokenProvider = () => Task.FromResult(token.AccessToken))
+                    .Build();
+
+                connection.On<string, string>("SendMessage", (_, json) =>
+                    this.ProcessNewMessage(station, direction, json)
+                        .ConfigureAwait(false)
+                        .GetAwaiter()
+                        .GetResult());
+
+                connection.Closed += async (e) =>
+                {
+                    if (e != null)
+                    {
+                        Log.Logger.Here().Warning(e, "SignalR connection was closed (unexpectedly?) as a result of an exception; recovering");
+                        await Task.Delay(new Random().Next(0, 50) * 100);
+                        await this.CreateHubConnection(tokenBrokerUrl, tokenValue, station, direction);
+                    }
+                    else
+                    {
+                        Log.Logger.Here().Warning("SignalR connection close event was triggered but the exception is null; expected?");
```

Comment:
This is happening for no reason,

In other words, the connection can be terminated without a valid exception -- potentially from the remote host or due to network issues.

https://i.imgur.com/QrouLsw.png

Seems like we should trigger a new connection in this case. We should test triggering a reconnect ourselves to see if Closed is called when it is the client that closes the connection.